File: src/main/java/c4/ast/C4Scanner.java
package c4.ast;

import c4.io.*;
import scala.Option;

import java_cup.runtime.Symbol;

import c4.util.Located;

public class C4Scanner implements java_cup.runtime.Scanner {
    private final SourcePhase7Reader reader;

    public C4Scanner(SourcePhase7Reader reader) {
        this.reader = reader;
    }

    @Override
    public Symbol next_token() throws Exception {
        Option<Located<Tok>> tok = reader.get();
        if (tok.isDefined()) {
            Located<Tok> located = tok.get();
            if (located.value() instanceof TokInteger) {
                TokInteger tokInt = (TokInteger) located.value();
                return new Symbol(CupSymbols.NUMBER, tokInt.n().bigInteger());
            } else if (located.value() instanceof Tok_$plus$) {
                return new Symbol(CupSymbols.PLUS);
            } else if (located.value() instanceof Tok_semicolon$) {
                return new Symbol(CupSymbols.SEMI);
            } else {
                throw new RuntimeException("not supported yet");
            }
        } else {
            return new Symbol(CupSymbols.EOF);
        }
    }
}

File: src/main/scala/c4/ast/C4Scanner.scala
package c4.ast

import java_cup.runtime.Scanner
import java_cup.runtime.Symbol

import c4.io._
import c4.io.SourcePhase7Reader


class C4Scanner(val reader: SourcePhase7Reader) extends Scanner {
  var scopeList: Seq[Set[String]] = Seq(Set.empty)

  def enterScope(): Unit = {
    scopeList = Set.empty[String] +: scopeList
  }

  def exitScope(): Unit = {
    scopeList = scopeList.tail
  }

  def isTypedefName(name: String): Boolean = {
    scopeList.exists(_.contains(name))
  }

  def addTypedefName(name: String): Unit = {
    (scopeList.head + name) +: scopeList.tail
  }

  @throws(classOf[java.lang.Exception])
  override def next_token(): Symbol = {
    reader.get() match {
      case None => new Symbol(CupSymbols.EOF)
      case Some(located) => located.value match {
        case tok: TokId =>
          if (isTypedefName(tok.id)) {
            new Symbol(CupSymbols.TYPEDEF_NAME, located)
          } else {
            new Symbol(CupSymbols.ID, located)
          }
        case tok: TokInteger    => new Symbol(tok.cupSymbol, located)
        case tok: TokFloat      => new Symbol(tok.cupSymbol, located)
        case tok: TokDouble     => new Symbol(tok.cupSymbol, located)
        case tok: TokLongDouble => new Symbol(tok.cupSymbol, located)
        case tok: TokChar       => new Symbol(tok.cupSymbol, located)
        case tok: TokWideChar   => new Symbol(tok.cupSymbol, located)
        case tok: TokStr        => new Symbol(tok.cupSymbol, located)
        case tok: TokWideStr    => new Symbol(tok.cupSymbol, located)
        case tok => new Symbol(tok.cupSymbol, located.loc)
      }
    }
  }
}
